cmd/kubelet/kubelet.go

```diff
@@ -21,108 +21,23 @@ limitations under the License.
 package main
 
 import (
-	"fmt"
 	"os"
 
-	"github.com/golang/glog"
-	"github.com/spf13/pflag"
-
-	utilfeature "k8s.io/apiserver/pkg/util/feature"
-	"k8s.io/apiserver/pkg/util/flag"
+	utilflag "k8s.io/apiserver/pkg/util/flag"
 	"k8s.io/apiserver/pkg/util/logs"
 	"k8s.io/kubernetes/cmd/kubelet/app"
-	"k8s.io/kubernetes/cmd/kubelet/app/options"
 	_ "k8s.io/kubernetes/pkg/client/metrics/prometheus" // for client metric registration
 	_ "k8s.io/kubernetes/pkg/version/prometheus"        // for version metric registration
-	"k8s.io/kubernetes/pkg/version/verflag"
 )
 
-func parseFlagSet(fs *pflag.FlagSet, args []string) error {
-	if err := fs.Parse(args); err != nil {
-		return err
-	}
-	fs.VisitAll(func(flag *pflag.Flag) {
-		glog.V(2).Infof("FLAG: --%s=%q", flag.Name, flag.Value)
-	})
-	return nil
-}
-
-func die(err error) {
-	fmt.Fprintf(os.Stderr, "error: %v\n", err)
-	os.Exit(1)
-}
-
 func main() {
-	fs := pflag.NewFlagSet(os.Args[0], pflag.ExitOnError)
-	// set the normalize func, similar to k8s.io/apiserver/pkg/util/flag/flags.go:InitFlags
-	fs.SetNormalizeFunc(flag.WordSepNormalizeFunc)
-	// explicitly add flags from libs that register global flags
-	options.AddGlobalFlags(fs)
+	command := app.NewKubeletCommand()
 
-	// register kubelet flags
-	kubeletFlags := options.NewKubeletFlags()
-	kubeletFlags.AddFlags(fs)
-
-	// register kubelet config flags
-	defaultConfig, err := options.NewKubeletConfiguration()
-	if err != nil {
-		die(err)
-	}
-	options.AddKubeletConfigFlags(fs, defaultConfig)
-
-	// parse flags
-	if err := parseFlagSet(fs, os.Args[1:]); err != nil {
-		die(err)
-	}
-
-	// initialize logging and defer flush
+	utilflag.InitFlags()
```

Comment:
Don't think you can do this - see https://github.com/kubernetes/kubernetes/blob/8a6bb3e1201ba7e1dcf0da1dea58fc72469d0c51/cmd/kube-proxy/proxy.go#L36-L38

Reply:
It seems to work ok for the scheduler: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/scheduler.go#L32

Reply:
> It seems to work ok for the scheduler: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/scheduler.go#L32-L34

I see the bug.  Registering global flags is naughty.  Fixed.

Reply:
As we discussed on slack, the scheduler is registering its flags globally into `pflag.CommandLine`. The kube-proxy registers its flags into the cobra command's `.Flags()`, which is why `utilflag.InitFlags()` doesn't work for kube-proxy but does for the scheduler. We need to fix the scheduler, finish converting all the executables to use cobra (#34732), then we can use InitFlags.